packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart

```diff
@@ -205,44 +170,89 @@ class AmplifyPushNotifications extends PushNotificationsPluginInterface {
-        alert: true,
-        sound: true,
-        badge: true,
+        alert: alert,
+        sound: badge,
+        badge: sound,
       ),
     );
   }
 
   @override
   Future<PushNotificationPermissionRequestStatus> getPermissionStatus() async {
-    final result = await _nativePlugin.getPermissionStatus();
-
-    if (result == 'Authorized') {
-      return PushNotificationPermissionRequestStatus.granted;
-    } else if (result == 'Denied') {
-      return PushNotificationPermissionRequestStatus.denied;
+    final result = await _hostApi.getPermissionStatus();
+
+    switch (result.status) {
+      case PermissionStatus.denied:
+        return PushNotificationPermissionRequestStatus.denied;
+      case PermissionStatus.granted:
+        return PushNotificationPermissionRequestStatus.granted;
+      case PermissionStatus.shouldRequestWithRationale:
+        return PushNotificationPermissionRequestStatus.shouldShowWithRationale;
+      case PermissionStatus.notRequested:
+        return PushNotificationPermissionRequestStatus.notRequested;
     }
-
-    return PushNotificationPermissionRequestStatus.shouldShowWithRationale;
   }
 
   @override
   PushNotificationMessage? get launchNotification => PushNotificationMessage();
 
   Future<PushNotificationMessage?> _getLaunchNotification() async {
-    await _nativePlugin.getLaunchNotification();
+    await _hostApi.getLaunchNotification();
 
     return PushNotificationMessage();
   }
 
   @override
   Future<int> getBadgeCount() {
-    return _nativePlugin.getBadgeCount();
+    return _hostApi.getBadgeCount();
   }
 
   @override
   Future<void> setBadgeCount(int badgeCount) async {
-    await _nativePlugin.setBadgeCount(badgeCount);
+    await _hostApi.setBadgeCount(badgeCount);
+  }
+}
+
+class _PushNotificationsFlutterApi implements PushNotificationsFlutterApi {
+  _PushNotificationsFlutterApi() {
+    PushNotificationsFlutterApi.setup(this);
+  }
+
+  final _onNotificationReceivedInBackgroundCallbacks =
+      <OnRemoteMessageCallback>[];
+  final _onLaunchNotificationOpenedCallbacks = <OnRemoteMessageCallback>[];
+
+  void registerOnReceivedInBackgroundCallback(
+    OnRemoteMessageCallback callback,
+  ) {
+    _onNotificationReceivedInBackgroundCallbacks.add(callback);
+  }
+
+  void registerOnLaunchNotificationOpenedCallback(
+    OnRemoteMessageCallback callback,
+  ) {
+    _onLaunchNotificationOpenedCallbacks.add(callback);
+  }
+
+  @override
+  Future<void> onNotificationReceivedInBackground(
+    Map<Object?, Object?> payload,
+  ) async {
+    await Future.wait(
+      _onNotificationReceivedInBackgroundCallbacks.map(
+        (callback) async {
+          await callback(PushNotificationMessage.fromJson(payload));
+        },
+      ),
+    );
+  }
+
+  @override
+  void onLaunchNotificationOpened(Map<Object?, Object?> payload) {
+    for (final callback in _onLaunchNotificationOpenedCallbacks) {
+      callback(PushNotificationMessage.fromJson(payload));
+    }
```

Comment:
Since it's intended to be used internally, there should be just one callback and then in theory you can just drop the reference to it too after calling it

Reply:
I'm thinking that even it's internal, we may need to add more than 1 callback to handle different use cases, e.g.
1. register a callback to set the launch notification for the `get launchNotification` getter
2. and one to execute the logic to record analytics event.

I've pushed an update to clear out the registered callbacks after execute them once (as onLaunchNotification should happen only once in one App life cycle)

Reply:
If a launch notification is defined as a notification which opens the app from killed, then I'm not I understand why the platform calls into the Dart code instead of, for example:

```dart
Future<void> configure(...) async {
  // ...
  _launchNotification = await _hostApi.getLaunchNotification(); 
  for (final callback in callbacks) {
    // ...
  }
}
```

Having this separate from `configure` doesn't seem to make sense.

Reply:
On iOS when a remote notification wakes the App in the background, it's the entire App wakes up, including developers Flutter App. `Amplify.configure()` would be executed at this point, `_launchNotification = await _hostApi.getLaunchNotification();` won't get the correct launch notification (the App is still running in the background, and there is no notification being tapped)